File: pythonlibrary/aibrd/generators/conflict_detector.py
from datetime import date
from ..llm.client import call_llm_json
from ..models.outputs import Conflict, BusinessRule

_SYSTEM = """You are a business analyst reviewing business rules for contradictions.
Find pairs of rules that directly conflict with each other.
Return JSON: {"conflicts": [{"rule_a": "...", "rule_b": "...", "description": "..."}]}"""


def detect_conflicts(rules: list[BusinessRule]) -> list[Conflict]:
    if len(rules) < 2:
        return []
    rules_text = "\n".join(f"{r.id}: {r.description}" for r in rules)
    raw = call_llm_json(rules_text, _SYSTEM)
    return [Conflict(**c) for c in raw.get("conflicts", [])]


def format_conflict_report(conflicts: list[Conflict]) -> str:
    today = date.today().isoformat()
    lines = ["# Conflict Report", f"_Generated: {today}_", ""]
    if not conflicts:
        lines.append("No conflicts detected.")
        return "\n".join(lines)

    lines.append("> Rule pairs that contradict each other. Resolve before development.")
    lines.append("")
    for i, c in enumerate(conflicts, 1):
        lines += [
            f"## Conflict {i}",
            f"- **Rule A:** {c.rule_a}",
            f"- **Rule B:** {c.rule_b}",
            f"- **Issue:** {c.description}",
            "",
        ]
    return "\n".join(lines)

File: src/core/generators/conflict_detector.ts
import * as vscode from 'vscode'
import { callLLMJson } from '../../llm/client'
import { Conflict, BusinessRule } from '../models/outputs'

const SYSTEM = `You are a business analyst reviewing business rules for contradictions.
Find pairs of rules that directly conflict with each other.
Return JSON: {
  "conflicts": [{
    "ruleA": "rule ID or description",
    "ruleB": "rule ID or description",
    "description": "why they conflict"
  }]
}`

export async function detectConflicts(
  rules: BusinessRule[],
  token?: vscode.CancellationToken
): Promise<Conflict[]> {
  if (rules.length < 2) return []

  const rulesText = rules
    .map(r => `${r.id}: ${r.description}`)
    .join('\n')

  const result = await callLLMJson<{ conflicts: Conflict[] }>(
    rulesText,
    SYSTEM,
    token
  )
  return result.conflicts
}

export function formatConflictReport(conflicts: Conflict[]): string {
  const lines: string[] = [
    '# Conflict Report',
    `_Generated: ${new Date().toISOString().split('T')[0]}_`,
    ''
  ]

  if (conflicts.length === 0) {
    lines.push('No conflicts detected.')
    return lines.join('\n')
  }

  lines.push('> The following rule pairs contradict each other. Resolve before development.', '')

  for (let i = 0; i < conflicts.length; i++) {
    const c = conflicts[i]
    lines.push(`## Conflict ${i + 1}`)
    lines.push(`- **Rule A:** ${c.ruleA}`)
    lines.push(`- **Rule B:** ${c.ruleB}`)
    lines.push(`- **Issue:** ${c.description}`)
    lines.push('')
  }

  return lines.join('\n')
}
